# Originally Minh's work

In [18]:
import numpy as np
import pandas as pd
import plotly.graph_objects as go
import plotly.express as px

### Cleaning

In [32]:
data = pd.read_csv('outage.csv').iloc[4:]
cols = list(data.iloc[0])[1:] # getting column names
data = data.drop(columns = "Major power outage events in the continental U.S.")
data.columns = cols
data = data.iloc[2:]
data

OBS  YEAR MONTH    U.S._STATE POSTAL.CODE NERC.REGION  \
6        1  2011     7     Minnesota          MN         MRO   
7        2  2014     5     Minnesota          MN         MRO   
8        3  2010    10     Minnesota          MN         MRO   
9        4  2012     6     Minnesota          MN         MRO   
10       5  2015     7     Minnesota          MN         MRO   
...    ...   ...   ...           ...         ...         ...   
1535  1530  2011    12  North Dakota          ND         MRO   
1536  1531  2006   NaN  North Dakota          ND         MRO   
1537  1532  2009     8  South Dakota          SD         RFC   
1538  1533  2009     8  South Dakota          SD         MRO   
1539  1534  2000   NaN        Alaska          AK        ASCC   

          CLIMATE.REGION ANOMALY.LEVEL CLIMATE.CATEGORY  \
6     East North Central          -0.3           normal   
7     East North Central          -0.1           normal   
8     East North Central          -1.5             cold   
9     East North Central          -0.1           normal   
10    East North Central           1.2             warm   
...                  ...           ...              ...   
1535  West North Central          -0.9             cold   
1536  West North Central           NaN              NaN   
1537  West North Central           0.5             warm   
1538  West North Central           0.5             warm   
1539                 NaN           NaN              NaN   

              OUTAGE.START.DATE  ... POPPCT_URBAN POPPCT_UC POPDEN_URBAN  \
6          Friday, July 1, 2011  ...        73.27     15.28         2279   
7          Sunday, May 11, 2014  ...        73.27     15.28         2279   
8     Tuesday, October 26, 2010  ...        73.27     15.28         2279   
9        Tuesday, June 19, 2012  ...        73.27     15.28         2279   
10      Saturday, July 18, 2015  ...        73.27     15.28         2279   
...                         ...  ...          ...       ...          ...   
1535  Tuesday, December 6, 2011  ...         59.9      19.9       2192.2   
1536                        NaN  ...         59.9      19.9       2192.2   
1537  Saturday, August 29, 2009  ...        56.65     26.73       2038.3   
1538  Saturday, August 29, 2009  ...        56.65     26.73       2038.3   
1539                        NaN  ...        66.02     21.56       1802.6   

     POPDEN_UC POPDEN_RURAL AREAPCT_URBAN AREAPCT_UC          PCT_LAND  \
6       1700.5         18.2          2.14        0.6  91.5926658691451   
7       1700.5         18.2          2.14        0.6  91.5926658691451   
8       1700.5         18.2          2.14        0.6  91.5926658691451   
9       1700.5         18.2          2.14        0.6  91.5926658691451   
10      1700.5         18.2          2.14        0.6  91.5926658691451   
...        ...          ...           ...        ...               ...   
1535    1868.2          3.9          0.27        0.1  97.5996492121418   
1536    1868.2          3.9          0.27        0.1  97.5996492121418   
1537    1905.4          4.7           0.3       0.15  98.3077441776026   
1538    1905.4          4.7           0.3       0.15  98.3077441776026   
1539      1276          0.4          0.05       0.02  85.7611544611833   

         PCT_WATER_TOT  PCT_WATER_INLAND  
6     8.40733413085488  5.47874298334407  
7     8.40733413085488  5.47874298334407  
8     8.40733413085488  5.47874298334407  
9     8.40733413085488  5.47874298334407  
10    8.40733413085488  5.47874298334407  
...                ...               ...  
1535  2.40176525502843  2.40176525502843  
1536  2.40176525502843  2.40176525502843  
1537  1.69225582239743  1.69225582239743  
1538  1.69225582239743  1.69225582239743  
1539  14.2388455388167  2.90118187392543  

[1534 rows x 56 columns]

In [20]:
data.shape

(1534, 56)

Data has 1534 rows and 56 columns

Let us consider which features interest us. 

Since we are focused on the duration of outages and it's temporal and spatial difference, we are most interested in columns with direct intuitive relation to these factors.

Hence, we will remove the columns: 'OBS', 'HURRICANE.NAMES', 'CUSTOMERS.AFFECTED', 'RES.PRICE', 'COM.PRICE', 'IND.PRICE',
       'TOTAL.PRICE', 'RES.SALES', 'COM.SALES', 'IND.SALES', 'TOTAL.SALES',
       'RES.PERCEN', 'COM.PERCEN', 'IND.PERCEN', 'RES.CUSTOMERS',
       'COM.CUSTOMERS', 'IND.CUSTOMERS', 'TOTAL.CUSTOMERS', 'RES.CUST.PCT',
       'COM.CUST.PCT', 'IND.CUST.PCT', 'PC.REALGSP.STATE', 'PC.REALGSP.USA',
       'PC.REALGSP.REL', 'PC.REALGSP.CHANGE', 'UTIL.REALGSP', 'TOTAL.REALGSP',
       'UTIL.CONTRI', 'PI.UTIL.OFUSA', 'POPULATION', 'POPPCT_URBAN',
       'POPPCT_UC', 'POPDEN_URBAN', 'POPDEN_UC', 'POPDEN_RURAL',
       'AREAPCT_URBAN', 'AREAPCT_UC', 'PCT_LAND', 'PCT_WATER_TOT',
       'PCT_WATER_INLAND'


<b>
Additionally
</b>
the detailed cause category will be monitored by us but not used in any EDA or analysis since it varies too much.

In [21]:
data = data.drop(columns = ['OBS', 'HURRICANE.NAMES', 'CUSTOMERS.AFFECTED', 'RES.PRICE', 'COM.PRICE', 'IND.PRICE',
       'TOTAL.PRICE', 'RES.SALES', 'COM.SALES', 'IND.SALES', 'TOTAL.SALES',
       'RES.PERCEN', 'COM.PERCEN', 'IND.PERCEN', 'RES.CUSTOMERS',
       'COM.CUSTOMERS', 'IND.CUSTOMERS', 'TOTAL.CUSTOMERS', 'RES.CUST.PCT',
       'COM.CUST.PCT', 'IND.CUST.PCT', 'PC.REALGSP.STATE', 'PC.REALGSP.USA',
       'PC.REALGSP.REL', 'PC.REALGSP.CHANGE', 'UTIL.REALGSP', 'TOTAL.REALGSP',
       'UTIL.CONTRI', 'PI.UTIL.OFUSA', 'POPULATION', 'POPPCT_URBAN',
       'POPPCT_UC', 'POPDEN_URBAN', 'POPDEN_UC', 'POPDEN_RURAL',
       'AREAPCT_URBAN', 'AREAPCT_UC', 'PCT_LAND', 'PCT_WATER_TOT',
       'PCT_WATER_INLAND', 'DEMAND.LOSS.MW', 'CAUSE.CATEGORY.DETAIL', 'U.S._STATE'])

Renaming for better intuitive use

In [22]:
data = data.rename(columns = {'YEAR': 'Year', 'MONTH': 'Month', 'POSTAL.CODE': 'State', 
                       'NERC.REGION': 'NERC Region', 'CLIMATE.REGION': 'Climate Region', 'ANOMALY.LEVEL': 'Anomaly Level', 'CLIMATE.CATEGORY': 'Climate',
                       'OUTAGE.START.DATE': 'Outage Start Date', 'OUTAGE.START.TIME': 'Outage Start Time', 'OUTAGE.RESTORATION.DATE': 'Outage Restoration Date',
                       'OUTAGE.RESTORATION.TIME': 'Outage Restoration Time', 'CAUSE.CATEGORY': 'Cause', 'OUTAGE.DURATION': 'Outage Duration'})

#### Addressing Column data types
Categorical(Non-Ordinal): Year, Month, State, NERC Region, Climate Region, Climate, Cause

Timestamp: Outage Start Date, Outate Start Time, Outage Restoration Date, Outage Restoration Time, Cause

Numerical(Discrete): Outage Duration

Numerical(Continuous): Anomaly level

In [23]:
data = data.sort_values(by = "Year")
categorical_cols = ['Year', 'Month', 'State', 'NERC Region', 'Climate Region', 'Climate', 'Cause']
for col in categorical_cols:
    data[col] = pd.Categorical(data[col])

timestamp_cols = ['Outage Start Date', 'Outage Start Time',
                  'Outage Restoration Date', 'Outage Restoration Time']
for col in timestamp_cols:
    data[col] = pd.to_datetime(data[col], errors='coerce')

data['Outage Start Time'] = pd.to_datetime(data['Outage Start Time'], format='%I:%M:%S %p', errors='coerce').dt.time
data['Outage Restoration Time'] = pd.to_datetime(data['Outage Restoration Time'], format='%I:%M:%S %p', errors='coerce').dt.time

/var/folders/q9/dv0qd5v93jv_7gz3d0y5d0xc0000gn/T/ipykernel_2715/554688027.py:9: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  data[col] = pd.to_datetime(data[col], errors='coerce')
/var/folders/q9/dv0qd5v93jv_7gz3d0y5d0xc0000gn/T/ipykernel_2715/554688027.py:9: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  data[col] = pd.to_datetime(data[col], errors='coerce')


In [24]:
data['Outage Duration'] = pd.to_numeric(data['Outage Duration'])
data['Anomaly Level'] = pd.to_numeric(data['Anomaly Level'])

In [25]:
data

Year Month State NERC Region Climate Region  Anomaly Level Climate  \
772  2000   NaN    NC        SERC      Southeast            NaN     NaN   
241  2000     3    TX         TRE          South           -1.1    cold   
245  2000   NaN    TX        FRCC          South            NaN     NaN   
307  2000     8    IN        ECAR        Central           -0.5    cold   
345  2000   NaN    AL        SERC      Southeast            NaN     NaN   
..    ...   ...   ...         ...            ...            ...     ...   
611  2016     7    PA         RFC      Northeast           -0.3  normal   
749  2016     7    NC        SERC      Southeast           -0.3  normal   
663  2016     5    UT        WECC      Southwest            0.6    warm   
661  2016     1    UT        WECC      Southwest            2.2    warm   
689  2016     2    UT        WECC      Southwest            2.0    warm   

    Outage Start Date Outage Start Time Outage Restoration Date  \
772               NaT               NaT                     NaT   
241        2000-03-18          16:00:00              2000-03-18   
245               NaT               NaT                     NaT   
307        2000-08-28          23:00:00                     NaT   
345               NaT               NaT                     NaT   
..                ...               ...                     ...   
611        2016-07-23          15:15:00              2016-07-23   
749        2016-07-08          20:50:00              2016-07-09   
663        2016-05-19          21:36:00              2016-05-20   
661        2016-01-17          12:00:00              2016-01-17   
689        2016-02-07          13:21:00              2016-02-07   

    Outage Restoration Time                          Cause  Outage Duration  
772                     NaT                 severe weather              NaN  
241                17:10:00  system operability disruption             70.0  
245                     NaT              equipment failure              NaN  
307                     NaT              equipment failure              NaN  
345                     NaT                 severe weather              NaN  
..                      ...                            ...              ...  
611                19:53:00  system operability disruption            278.0  
749                19:25:00                 severe weather           1355.0  
663                01:00:00  system operability disruption            204.0  
661                13:00:00             intentional attack             60.0  
689                13:42:00             intentional attack             21.0  

[1534 rows x 13 columns]

#### Missing values

In [26]:
missing = data.isna().sum(axis = 0)
missing_cols = {row: int(missing[row]) for row in missing.index if missing[row] > 0}
missing_cols

{'Month': 9,
 'Climate Region': 6,
 'Anomaly Level': 9,
 'Climate': 9,
 'Outage Start Date': 9,
 'Outage Start Time': 9,
 'Outage Restoration Date': 58,
 'Outage Restoration Time': 58,
 'Outage Duration': 58}

In [27]:
data[data['Year'] == '2000']

Year Month State NERC Region Climate Region  Anomaly Level Climate  \
772   2000   NaN    NC        SERC      Southeast            NaN     NaN   
241   2000     3    TX         TRE          South           -1.1    cold   
245   2000   NaN    TX        FRCC          South            NaN     NaN   
307   2000     8    IN        ECAR        Central           -0.5    cold   
345   2000   NaN    AL        SERC      Southeast            NaN     NaN   
346   2000    12    AL        SERC      Southeast           -0.8    cold   
347   2000     8    AL        SERC      Southeast           -0.5    cold   
356   2000     5    IL        SERC        Central           -0.7    cold   
371   2000   NaN    IL        SERC        Central            NaN     NaN   
387   2000     8    IL        SERC        Central           -0.5    cold   
510   2000     6    AZ        WECC      Southwest           -0.7    cold   
236   2000     5    TX         SPP          South           -0.7    cold   
518   2000     5    AZ        WECC      Southwest           -0.7    cold   
737   2000     5    NC        SERC      Southeast           -0.7    cold   
751   2000     8    NC        SERC      Southeast           -0.5    cold   
816   2000     1    SC        SERC      Southeast           -1.6    cold   
818   2000     1    SC        SERC      Southeast           -1.6    cold   
893   2000   NaN    DE         RFC      Northeast            NaN     NaN   
929   2000     5    NY        NPCC      Northeast           -0.7    cold   
1219  2000     6    CA        WECC           West           -0.7    cold   
1324  2000   NaN    VA        SERC      Southeast            NaN     NaN   
1365  2000     3    NM        WECC      Southwest           -1.1    cold   
1469  2000     3    ME        NPCC      Northeast           -1.1    cold   
695   2000     6    OH        ECAR        Central           -0.7    cold   
193   2000     5    TX         TRE          South           -0.7    cold   
1539  2000   NaN    AK        ASCC            NaN            NaN     NaN   

     Outage Start Date Outage Start Time Outage Restoration Date  \
772                NaT               NaT                     NaT   
241         2000-03-18          16:00:00              2000-03-18   
245                NaT               NaT                     NaT   
307         2000-08-28          23:00:00                     NaT   
345                NaT               NaT                     NaT   
346         2000-12-16          11:36:00              2000-12-18   
347         2000-08-10          21:30:00              2000-08-11   
356         2000-05-18          18:00:00                     NaT   
371                NaT               NaT                     NaT   
387         2000-08-06          16:00:00              2000-08-07   
510         2000-06-14          15:54:00              2000-06-14   
236         2000-05-24          10:15:00              2000-05-25   
518         2000-05-31          01:15:00                     NaT   
737         2000-05-25          10:00:00              2000-06-02   
751         2000-08-18          18:30:00              2000-08-20   
816         2000-01-23          08:00:00              2000-01-28   
818         2000-01-29          22:00:00              2000-02-03   
893                NaT               NaT                     NaT   
929         2000-05-09          11:39:00              2000-05-09   
1219        2000-06-14          13:13:00                     NaT   
1324               NaT               NaT                     NaT   
1365        2000-03-18          19:08:00              2000-03-18   
1469        2000-03-14          21:06:00                     NaT   
695         2000-06-14          15:45:00                     NaT   
193         2000-05-02          04:00:00              2000-05-02   
1539               NaT               NaT                     NaT   

     Outage Restoration Time                          Cause  Outage Duration  
772                      NaT         

### EDA

In [28]:
year_duration = go.Figure()

for year in data['Year'].unique():
    year_data = data[data['Year'] == year]
    year_duration.add_trace(go.Scatter(
        x=year_data['State'],
        y=year_data['Outage Duration'],
        mode='markers',
        name=year,
        marker=dict(size=6),
        opacity=0.7
    ))

year_duration.update_layout(
    title='Outage Duration by State (Colored by Year)',
    xaxis_title='State',
    yaxis_title='Outage Duration',
    showlegend=True
)

year_duration.show()

In [29]:
month_duration = go.Figure()

for month in data['Month'].unique():
    month_data = data[data['Month'] == month]
    month_duration.add_trace(go.Scatter(
        x=month_data['State'],
        y=month_data['Outage Duration'],
        mode='markers',
        name=f'Month {month}',
        marker=dict(size=6),
        opacity=0.7
    ))

month_duration.update_layout(
    title='Outage Duration by State (Colored by Month)',
    xaxis_title='State',
    yaxis_title='Outage Duration',
    showlegend=True
)

month_duration.show()


In [30]:
climate_fig = go.Figure()

for region in data['Climate Region'].unique():
    region_data = data[data['Climate Region'] == region]
    climate_fig.add_trace(go.Scatter(
        x=region_data['Year'],
        y=region_data['Outage Duration'],
        mode='markers',
        name=region,
        marker=dict(size=6),
        opacity=0.7
    ))

climate_fig.update_layout(
    title='Outage Duration vs Year (Colored by Climate Region)',
    xaxis_title='Year',
    yaxis_title='Outage Duration',
    showlegend=True
)

climate_fig.show()